src/logger.js

```diff
@@ -1,38 +1,121 @@
+const dotenv = require('dotenv');
+
 /* istanbul ignore file */
+const { LOGLEVELS, structuredLogger } = require('@rudderstack/integrations-lib');
 
-const levelDebug = 0; // Most verbose logging level
-const levelInfo = 1; // Logs about state of the application
-const levelWarn = 2; // Logs about warnings which dont immediately halt the application
-const levelError = 3; // Logs about errors which dont immediately halt the application
 // any value greater than levelError will work as levelNone
+const loggerImpl = process.env.LOGGER_IMPL ?? 'winston';
+
+let logLevel = process.env.LOG_LEVEL ?? 'error';
 
-let logLevel = process.env.LOG_LEVEL ? parseInt(process.env.LOG_LEVEL, 10) : levelInfo;
+const logger = structuredLogger({ level: logLevel });
+
+const getLogger = () => {
+  return loggerImpl === 'winston' ? logger : console;
+};
 
 const setLogLevel = (level) => {
+  const logger = getLogger();
   logLevel = level || logLevel;
+  logger?.setLogLevel(logLevel);
+};
+
+/**
+ * obtains the metadata for logging
+ *
+ * @param {*} metadata
+ * @returns { destinationId:string, sourceId:string, workspaceId: string, destType:string, module:string, implementation:string, feature:string }
+ */
+const getLogMetadata = (metadata) => {
+  let reqMeta = metadata;
+  if (Array.isArray(metadata)) {
+    [reqMeta] = metadata;
+  }
+  return {
+    ...(reqMeta?.destinationId && { destinationId: reqMeta.destinationId }),
+    ...(reqMeta?.sourceId && { sourceId: reqMeta.sourceId }),
+    ...(reqMeta?.workspaceId && { workspaceId: reqMeta.workspaceId }),
+    ...(reqMeta?.destType && { destType: reqMeta.destType }),
+    ...(reqMeta?.module && { module: reqMeta.module }),
+    ...(reqMeta?.implementation && { implementation: reqMeta.implementation }),
+    ...(reqMeta?.feature && { feature: reqMeta.feature }),
+  };
+};
+
+const log = (logMethod, args) => {
+  const [message, logInfo, ...otherArgs] = args;
+  if (logInfo) {
+    const { metadata, ...otherLogInfoArgs } = logInfo;
+    if (Array.isArray(metadata)) {
+      metadata.forEach((m) => {
+        logMethod(
+          message,
+          {
+            ...getLogMetadata(m),
+            ...otherLogInfoArgs,
+          },
+          ...otherArgs,
+        );
+      });
+      return;
+    }
+    logMethod(
+      message,
+      {
+        ...getLogMetadata(metadata),
+        ...otherLogInfoArgs,
+      },
+      ...otherArgs,
+    );
+    return;
+  }
+  logMethod(message);
 };
 
```

Comment:
Add comment mentioning debug/info/warn/error implementation below are applicable on for `console` logger